### DISCUSSION TITLE
open a view and close another: ActionHandler

### DISCUSSION THREAD

**Discussion by Adem-Kri** [eclipse-glsp/glsp#838] [General] [answered]
Hello,
I created a button inside my glsp view similar to the button overlay showcase existed in the documentation.
So, I have a glsp view  and a button and I want to implement an actionHandler that closes the opened view and open a theia extension,
this is the button source code, inside `Switch()` function is where the process will be implemented
```
containerElement.appendChild(
        this.createButton(
          "btn_switch_diagram",
          "Switch to table view",
          "arrow-both",
          SwitchAction.Switch()
        )
      );
```
I have made some researches and came out with this: I need to implement the `openerService `and `widgetOpenerOptions`,
but I am still a beginner with theia and didn't know how to do it, am seeking for your help and thank you in advance.

**Comment by planger** [comment, marked as answer]
Hi,

this is more a Theia-related question, but I'll try to outline what is involved for this task.

In Theia there is an [opener service](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/opener-service.ts#L94), which you can use to open widgets or editors based on opener options. You can get the opener service injected from Theia's dependency injection container and invoke it as follows (note that you can only [get things injected in classes that are instantiated by the Theia dependency injection itself](https://theia-ide.org/docs/services_and_contributions#dependency-injection-di)):

```typescript
import { OpenerService, open } from '@theia/core/lib/browser';
...
@inject(OpenerService) protected readonly openerService: OpenerService;
....
open(this.openerService, uri, openerOptions);
```

The opener service will then query all [open handlers](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/opener-service.ts#L28) that are registered via the contribution point `bind(OpenHandler).to(AnOpenHandler)`. Each open handler needs to return a priority for a provided URI and opener options. The one returning the highest priority will be used by default, others may only show up in a list, e.g. from the file explorer "Open With" menu.

A typical base class for a custom open handler is e.g. [`NavigatableWidgetOpenHandler`](https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/navigatable.ts).

So in your case it depends if you want to open (a) a custom widget or (b) a widget / editor for which there already is an open handler.
In case of (a) you need to register your custom open handler which will return the highest priority for a specific URI and opener options that you use in your `open` call from your switch action. This custom open handler would open your custom editor / widget. This implementation however is typically reused by subclassing e.g. NavigatableWidgetOpenHandler and just pointing to your widget ID, so that you just need to implement the `canHandle` method in your custom open handler. For (b) it is enough to just invoke the right `open` call from your switch action.
